crawler.c

```diff
@@ -349,6 +349,58 @@ static void lru_crawler_class_done(int i) {
         active_crawler_mod.mod->doneclass(&active_crawler_mod, i);
 }
 
+static void item_crawl_hash(void) {
+    // get iterator from assoc. can hang for a long time.
+    // - blocks hash expansion
+    void *iter = assoc_get_iterator();
+    int crawls_persleep = settings.crawls_persleep;
+    item *it = NULL;
+
+    // loop while iterator returns something
+    // - iterator func handles bucket-walking
+    // - iterator returns with bucket locked.
+    while ((it = assoc_iterate(iter)) != NULL) {
+        /* Get memory from bipbuf, if client has no space, flush. */
+        if (active_crawler_mod.c.c != NULL) {
+            int ret = lru_crawler_client_getbuf(&active_crawler_mod.c);
+            if (ret != 0) {
+                // fail out and finalize.
+                break;
+            }
+        } else if (active_crawler_mod.mod->needs_client) {
+            // fail out and finalize.
+            break;
+        }
+
+        // double check that the item isn't in a transitional state.
+        if (refcount_incr(it) < 2) {
+            refcount_decr(it);
+            continue;
+        }
+
+        // FIXME: missing hv and i are fine for metadump eval, but not fine
+        // for expire eval.
+        active_crawler_mod.mod->eval(&active_crawler_mod, it, 0, 0);
+
+        // - sleep bits from orig loop
+        if (crawls_persleep-- <= 0 && settings.lru_crawler_sleep) {
```

Comment:
Wouldn't this hold up a bucket's lock for longer than necessary?

If we need to adhere to the sleep, maybe we can do so between buckets instead of 'crawls_persleep' for the hash table walk? That would require some changes to assoc_iterate() to return identifiably between buckets.

Reply:
good catch! that one totally slipped. You're probably right but I'll give it a think. The iterator never returns without a lock, but I could hand it a between-unlocks callback during init without making the interface too gross.